kolibri/core/tasks/api.py

```diff
@@ -207,8 +244,60 @@ def list(self, request):
         return Response(jobs_response)
 
     def create(self, request):
-        # unimplemented. Call out to the task-specific APIs for now.
-        pass
+        """
+        Enqueue a task for async processing.
+
+        API endpoint:
+            POST /api/tasks/
+
+        Request payload parameters:
+            - `task` (required): a string representing the dotted path to task function.
+            - all other key value pairs are passed to the validator if the
+              task function has one otherwise they are passed to the task function itself
+              as keyword args.
+
+        Keep in mind:
+            If a task function has a validator then dict returned by the validator
+            is passed to the task function as keyword args.
+
+            The validator can add `extra_metadata` in the returning dict to set `extra_metadata`
+            in the enqueued task.
+        """
+        request_data_list = self.validate_create_req_data(request)
+
+        enqueued_jobs_response = []
+
+        # Once we have validated all the tasks, we are good to go!
+        for request_data in request_data_list:
+
+            funcstr = request_data.get("task")
+            registered_job = JobRegistry.REGISTERED_JOBS[funcstr]
+
+            # Run validator with request and request_data as its argument
+            if registered_job.validator is not None:
+                try:
+                    validator_result = registered_job.validator(request, request_data)
+                except Exception as e:
+                    raise e
+
+                if not isinstance(validator_result, dict):
+                    raise serializers.ValidationError("Validator must return a dict.")
+
+                extra_metadata = validator_result.get("extra_metadata")
+                if extra_metadata is not None and not isinstance(extra_metadata, dict):
+                    raise serializers.ValidationError(
+                        "In the dict returned by validator, 'extra_metadata' must be a dict."
+                    )
+
+                request_data = validator_result
+
+            job_id = registered_job.enqueue(**request_data)
```

Comment:
`task` key should be removed from the `request_data`. With the current implementation the task function is forced to have an unused arg called `task`

Reply:
Given that it's already been validated to have `"task"` in the data, can use:
```
funcstr = request_data.pop("task")
```

Above - using pop instead of get to address this!

Reply:
Yes, this will be fixed in the upcoming PR that aims to decouple some tasks to their respective places.

Reply:
hi , @vkWeb  I'm using the current implementation for a new feature. If you plan to do changes on the api it's important to know it.
There are a some things I have missed using this new tasks api:
1. Possibility to select a different queue (only normal or high are available)
2. A JS api (I've done my own but having a general to inherit would be great)
3. Possibility to change the permissions of some of the TaskViewSet functions. I have had to create a new class inheriting it just to be able to change the permissions to see the list of tasks.

In any case, none of these things are fundamental and I've found workarounds for them, but having them would make a dev life easier. The general work done on the tasks is great and being able to register new tasks from a plugin is awesome. I have had to delete code from a previous implementation and that's my favourite refactoring ;)

Reply:
Hi sir @jredrejo. I'm reallyyy happy to hear that! 

Right now, the Tasks API and the tasks backend are under active refactoring. In the span of next one month we can expect our tasks backend to give a much more delightful developer experience. I'll keep you updated on the developments via Slack!     

> Possibility to select a different queue (only normal or high are available).
> I have read in the specs the possility to use groupName to group tasks but it's not implemented as far as I have seen.

After implementing single worker pool, we'll have the ability to give tasks any queue we want, later on we can fetch/delete tasks by any particular queue. Moreover, `Queue` interface will be removed, queue will just be a string column on the job storage table. Then, `high` and `normal` priority levels will be used to decide which task should be fetched first by the workers.   

I've planned to implement single worker pool after I shift tasks from the `kolibri.core.tasks.api` monolith to their respective django apps. Once shifting is complete, we'll get the freedom to mould our tasks backend in whatever way we want without disturbing the working of tasks itself.

> A JS api (I've done my own but having a general to inherit would be great)

A JS API...? I'm not sure I understand..?

> Possibility to change the permissions of some of the TaskViewSet functions. I have had to create a new class inheriting it just to be able to change the permissions to see the list of tasks.

The `register_task` decorator has a `permission_classes` argument that'll allow flexible setting of permissions for any particular task. I would request you try not to inherit TaskViewSet, let us discuss the problem to see the pain that's forcing us to inherit.

Reply:
@jredrejo just to flag, all the upcoming PRs related to tasks backend will target https://github.com/learningequality/kolibri/tree/tasks_api_refactor branch so that it don't cause trouble in the 0.15 release.

Reply:
Hello again @vkWeb  I have been working on this in this PR  #8185 so all the code I'll mention here is there if you want to take a look

> > A JS api (I've done my own but having a general to inherit would be great)
> 
> A JS API...? I'm not sure I understand..?
> 

Something in https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/core/assets/src/api-resources/task.js to make all this new taks endpoints easier to be used from JS. I had to add in _kolibri/plugins/setup_wizard/assets/src/api.js_ some functions to handle it  (`SetupSoUDTasksResource` object)

> > Possibility to change the permissions of some of the TaskViewSet functions. I have had to create a new class inheriting it just to be able to change the permissions to see the list of tasks.
> 
> The `register_task` decorator has a `permission_classes` argument that'll allow flexible setting of permissions for any particular task. I would request you try not to inherit TaskViewSet, let us discuss the problem to see the pain that's forcing us to inherit.

If you take a look at #8185 you can find in _kolibri/plugins/setup_wizard/tasks.py_  the `startprovisionsoud` with its own permission_classes as you say. So far so good. 
The problem comes when the frontend needs to pull the tasks list to know if this task is in process. This code is being executed when provisioning kolibri after been installled, so there are not users yet and the permissions in `startprovisionsoud` work fine, but there's no way to get the list of the tasks from the frontend because `TasksViewSet` has a different permissions for `list`. 
I had to creeate  the class `SetupWizardSoUDTaskView(TasksViewSet) `at _kolibri/plugins/setup_wizard/api.py_ in #8185 to be able to use `SetupSoUDTasksResource.fetchCollection`  in _kolibri/plugins/setup_wizard/assets/src/views/importLODUsers/LoadingTaskPage.vue_ . If you know of a better solution to poll the tasks without inheriting `TaskViewSet` I am all ears :)

Reply:
> Something in https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/core/assets/src/api-resources/task.js to make all this new taks endpoints easier to be used from JS. I had to add in kolibri/plugins/setup_wizard/assets/src/api.js some functions to handle it (SetupSoUDTasksResource object).

I have not touched the frontend code at all because my all focus has been on first refactoring the tasks backend.

> The problem comes when the frontend needs to pull the tasks list to know if this task is in process. This code is being executed when provisioning kolibri after been installled, so there are not users yet and the permissions in startprovisionsoud work fine, but there's no way to get the list of the tasks from the frontend because TasksViewSet has a different permissions for list.

Oh, I get it. This should be handled on the TasksViewSet level, we must not inherit it. 

Ummm, looks like @rtibbles initial suggestion will prove helpful now. His suggestion was to filter the tasks based on whether user has permission for it or not, if the user do have permission then that task gets included in the response.  

So, in our case of `startprovisionsoud`, if the user sending API request to `list` has permission `[IsSuperuser | NotProvisionedCanPost | LODUserHasSyncPermissions]` the `startprovisionsoud` will be included in the response list. 

This feels like a good general solution to this problem. What do you think sir @jredrejo?

Reply:
> 
> Ummm, looks like @rtibbles initial suggestion will prove helpful now. His suggestion was to filter the tasks based on whether user has permission for it or not, if the user do have permission then that task gets included in the response.
> 
> So, in our case of `startprovisionsoud`, if the user sending API request to `list` has permission `[IsSuperuser | NotProvisionedCanPost | LODUserHasSyncPermissions]` the `startprovisionsoud` will be included in the response list.
> 
> This feels like a good general solution to this problem. What do you think sir @jredrejo?

it sounds perfect !

Reply:
If you say I'll send a PR right away...?